### PR TITLE
feat: upgraded typescript-eslint packages to v2.0

### PR DESCRIPTION
**Pull request by toddbluhm**
Note: this was branched from https://github.com/standard/eslint-config-standard-with-typescript/pull/134

Should resolve: https://github.com/standard/eslint-config-standard-with-typescript/issues/122

This should contain no breaking changes as far as rule changes go. There may be breaking changes from the `typescript-eslint/*^2.0.0` updates.

- Updated to use `@typescript-eslint/parser@^2.0.0`
- Updated to use `@typescript-eslint/eslint-plugin@^2.0.0`
- Updated `@typescript-eslint/array-type` to use the new options format, no
rule change was needed
- Replaced `@typescript-eslint/no-angle-bracket-type-assertion` and
`@typescript-eslint/no-object-literal-type-assertion` with new merged
version `@typescript-eslint/consistent-type-assertions` that should hold
the same defaults as before
- Added `commitlint.config.js` and `fixture.js` to `.eslintignore` file as
the new `typescript-eslint` wants them included in the project `tsconfig.js`
file which led to bad results (perhaps there is a better way around this?)
- Updated test case to support the new rule names and settings
- Updated `package.json` to include the new minimum version of `2.0.0` for
`typescript-eslint` packages

### CODE REVIEW

**Comment by mightyiam** [comment]
Hi, @toddbluhm.

Thank you for this PR.

I know that the plugin dependency can be updated without actually using the rules. But I made it a policy to review and decide on new rules when updating.

I have just submitted #137. After that is resolved and merged I will review this PR.

Would you be interested in reviewing the rule decisions in #137? That would help.

**Comment by toddbluhm** [comment]
Just wanted to add, based on your work with this shared config and based on this PR, I went ahead and created a utility based on `standard-engine` that works identical to `standard`, but for typescript. Its called [`ts-standard`](https://github.com/toddbluhm/ts-standard). 

Feel free to check it out. I have v1 already published to npm and it seems to work well. I have started using it on one of my other OSS projects already with good success. I am sure there are probably some edge cases I am missing and some tweaking that will need to happen, but I think it is a solid start.

File: src/index.test.ts [hunk 1]
```diff
@@ -18,9 +18,17 @@ test('export', (t): void => {
...
-          '@typescript-eslint/array-type': ['error', 'array-simple'],
+          '@typescript-eslint/array-type': ['error', { default: 'array-simple' }],
           '@typescript-eslint/camelcase': ['error', { properties: 'never' }],
+          '@typescript-eslint/consistent-type-assertions': [
+            'error',
+            {
+              assertionStyle: 'as',
+              objectLiteralTypeAssertions: 'never'
+            }
+          ],
```
Comment (mightyiam) [inline src/index.test.ts]: https://github.com/typescript-eslint/typescript-eslint/commit/92e98de?short_path=7c37156#diff-7c37156990de509cced1ad61a781c425
:+1:

File: src/index.test.ts [hunk 2]
```diff
@@ -48,17 +56,16 @@ test('export', (t): void => {
               singleline: { delimiter: 'comma', requireLast: false }
             }
           ],
-          '@typescript-eslint/no-angle-bracket-type-assertion': 'error',
```
Comment (mightyiam) [inline src/index.test.ts]: https://github.com/typescript-eslint/typescript-eslint/commit/92e98de?short_path=7c37156#diff-7c37156990de509cced1ad61a781c425
:+1:

File: src/index.test.ts [hunk 3]
```diff
@@ -48,17 +56,16 @@ test('export', (t): void => {
...
           '@typescript-eslint/no-empty-function': 'error',
           '@typescript-eslint/no-empty-interface': 'error',
           '@typescript-eslint/no-extraneous-class': 'error',
           '@typescript-eslint/no-floating-promises': 'error',
           '@typescript-eslint/no-for-in-array': 'error',
           '@typescript-eslint/no-misused-new': 'error',
+          '@typescript-eslint/no-misused-promises': 'error',
           '@typescript-eslint/no-namespace': 'error',
           '@typescript-eslint/no-non-null-assertion': 'error',
-          '@typescript-eslint/no-object-literal-type-assertion': 'error',
```
Comment (mightyiam) [inline src/index.test.ts]: https://github.com/typescript-eslint/typescript-eslint/commit/92e98de?short_path=7c37156#diff-7c37156990de509cced1ad61a781c425
:+1:

File: src/index.ts [hunk 4]
```diff
@@ -39,10 +40,18 @@ export = {
...
         '@typescript-eslint/adjacent-overload-signatures': 'error',
-        '@typescript-eslint/array-type': ['error', 'array-simple'],
+        '@typescript-eslint/array-type': ['error', { default: 'array-simple' }],
+        '@typescript-eslint/consistent-type-assertions': [
+          'error',
+          {
+            assertionStyle: 'as',
+            objectLiteralTypeAssertions: 'never'
+          }
+        ],
```
Comment (mightyiam) [inline src/index.ts]: https://github.com/typescript-eslint/typescript-eslint/commit/92e98de?short_path=7c37156#diff-7c37156990de509cced1ad61a781c425

:+1:

File: src/index.ts [hunk 5]
```diff
@@ -56,15 +65,14 @@ export = {
             singleline: { delimiter: 'comma', requireLast: false }
           }
         ],
-        '@typescript-eslint/no-angle-bracket-type-assertion': 'error',
```
Comment (mightyiam) [inline src/index.ts]: https://github.com/typescript-eslint/typescript-eslint/commit/92e98de?short_path=7c37156#diff-7c37156990de509cced1ad61a781c425
:+1:

File: src/index.ts [hunk 6]
```diff
@@ -56,15 +65,14 @@ export = {
...
         '@typescript-eslint/no-empty-interface': 'error',
         '@typescript-eslint/no-extraneous-class': 'error',
         '@typescript-eslint/no-floating-promises': 'error',
         '@typescript-eslint/no-for-in-array': 'error',
         '@typescript-eslint/no-misused-new': 'error',
+        '@typescript-eslint/no-misused-promises': 'error',
         '@typescript-eslint/no-namespace': 'error',
         '@typescript-eslint/no-non-null-assertion': 'error',
-        '@typescript-eslint/no-object-literal-type-assertion': 'error',
```
Comment (mightyiam) [inline src/index.ts]: https://github.com/typescript-eslint/typescript-eslint/commit/92e98de?short_path=7c37156#diff-7c37156990de509cced1ad61a781c425
:+1:

File: src/index.test.ts [hunk 7]
```diff
@@ -18,9 +18,17 @@ test('export', (t): void => {
           'no-unused-vars': 'off',
           'no-use-before-define': 'off',
           'no-useless-constructor': 'off',
+          'require-await': 'off',
           '@typescript-eslint/adjacent-overload-signatures': 'error',
-          '@typescript-eslint/array-type': ['error', 'array-simple'],
+          '@typescript-eslint/array-type': ['error', { default: 'array-simple' }],
```
Comment (mightyiam) [inline src/index.test.ts]: https://github.com/typescript-eslint/typescript-eslint/commit/1389393?short_path=8efdecc#diff-8efdecc7b1999c1e9a6b0790e12fddaa :+1:

File: src/index.ts [hunk 8]
```diff
@@ -39,10 +40,18 @@ export = {
         '@typescript-eslint/no-unused-vars': ['error', { args: 'none' }],
         '@typescript-eslint/no-use-before-define': ['error', { functions: false, classes: false, variables: false, typedefs: false }],
         '@typescript-eslint/no-useless-constructor': 'error',
+        '@typescript-eslint/require-await': 'error',
 
         // New Typescript-only rules:
         '@typescript-eslint/adjacent-overload-signatures': 'error',
-        '@typescript-eslint/array-type': ['error', 'array-simple'],
+        '@typescript-eslint/array-type': ['error', { default: 'array-simple' }],
```
Comment (mightyiam) [inline src/index.ts]: https://github.com/typescript-eslint/typescript-eslint/commit/1389393?short_path=8efdecc#diff-8efdecc7b1999c1e9a6b0790e12fddaa :+1:

File: src/index.test.ts [hunk 9]
```diff
@@ -18,9 +18,17 @@ test('export', (t): void => {
           'no-unused-vars': 'off',
           'no-use-before-define': 'off',
           'no-useless-constructor': 'off',
+          'require-await': 'off',
```
Comment (mightyiam) [inline src/index.test.ts]: This shouldn't be here, by now. Please rebase against `master`.

File: src/index.test.ts [hunk 10]
```diff
@@ -48,17 +56,16 @@ test('export', (t): void => {
...
           ],
-          '@typescript-eslint/no-angle-bracket-type-assertion': 'error',
           '@typescript-eslint/no-array-constructor': 'error',
           '@typescript-eslint/no-empty-function': 'error',
           '@typescript-eslint/no-empty-interface': 'error',
           '@typescript-eslint/no-extraneous-class': 'error',
           '@typescript-eslint/no-floating-promises': 'error',
           '@typescript-eslint/no-for-in-array': 'error',
           '@typescript-eslint/no-misused-new': 'error',
+          '@typescript-eslint/no-misused-promises': 'error',
```
Comment (mightyiam) [inline src/index.test.ts]: This is also already in `master`. Please rebase.

File: .eslintignore [hunk 11]
```diff
@@ -1 +1,3 @@
 lib
+commitlint.config.js
+fixture.js
```
[the remark this answers is not in the document]
Comment (toddbluhm) [reply .eslintignore]: So this is an interesting issue. I discovered the `tsconfig.eslint.json` thing yesterday and forgot to update this PR because I was still running into issues with `fixture.js`. 

Turns out, if you have two files with the same name but just different `.ts/.js` endings, typescript will only pickup/include the `.ts` version of the file. It makes sense because typescript is about compiling to js, so it does not want naming conflicts. Having said that, how do you want to go about solving this issue? We have two files `fixture.ts` and `fixture.js` in the same folder. I am guessing we should rename one of them? Is there a reason we are keeping `fixture.js` around since it is all commented out code anyways?

[the remark this answers is not in the document]
Comment (mightyiam) [reply .eslintignore]: Those fixture files helped me when typescript-eslint was much more buggy. So I don't care about them much.

The picking up of only the `.ts` file... why would it occur when we have `outDir` specified?

In any case, feel free to remove any of them.

And I feel that it would be prudent to report this issue to typescript-eslint. Could you do that, please?

[the remark this answers is not in the document]
Comment (toddbluhm) [reply .eslintignore]: Reported here: https://github.com/typescript-eslint/typescript-eslint/issues/890
I removed the `fixture.js` one since that seemed the least useful.

**Comment by toddbluhm** [comment]
@mightyiam ready for re-review on this.

File: tsconfig.eslint.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "allowJs": true,
+    "checkJs": true
+  },
+  "include": [
+    "src/**/*",
+    "*.ts",
+    "*.js"
+  ]
```
Comment (mightyiam) [inline tsconfig.eslint.json]: This seems to be enough:
```json
  "include": [
    "**/*"
  ]
```

I'll merge and change.

**Comment by mightyiam** [comment]
Thank you, @toddbluhm !
